docs/config-merge-logic.md

```diff
@@ -0,0 +1,609 @@
+  in {119, 120, 121, 122, 123}, ignoring any config message with a seqno of 118 or earlier.
+
+  This choice is configurable: higher values allow for a larger conflict window before data is
+  dropped, while lower values reduce the overhead of the diffs that are included in messages to
+  handle conflicts.
+
+- "ignore included": if a client sees two config messages with different seqno values and the
+  smaller one is contained within the diff of the larger one then the smaller one is stale and will
+  be ignored.
+
+- invalid messages: messages that are unparseable or fail some of the other requirements shall be
+  ignored; for example:
+  - duplicates or improperly sorted data inside a application data list
+  - missing required fields
+  - keys before `"#"` in the config message
+  - improperly sorted lagged config diffs
+  - etc.
+
+## Non-conflicting updates
+
+If a client is making a configuration change and there is only a single current valid config message
+(i.e. no conflicts to resolve) then the update procedure is straightforward:
+
+1. A new seqno is assigned that is 1 higher than the current largest seqno.
+2. The lagged config diffs from the existing config message are copied into the new message,
+   omitting config messages with seqno values less than or equal to (newseqno-N).  The client should
+   ensure the values are sorted by seqno, hash.  (They should be sorted already, but clients should
+   not propagate wrongly sorted lists if they encounter one).
+3. The current config message's diff is appended to the lagged config diff (using the message's
+   seqno value and calculating a hash of the message to complete the 3-tuple).
+4. A current diff is constructed for any values being assigned, and stored under the `"="` key of
+   the config message.
+
+## Conflict resolution
+
+Conflict resolution logic kicks in if, after removing ignored messages from consideration, there are
+still multiple valid config messages: these messages must be merged and the merge update published.
+
+Merging is performed as follows:
+
+1. A new seqno is assigned that is 1 higher than the current largest seqno.
+2. All messages are sorted by their seqno (first) and hash (second).
+3. The "current data" is initialized from the full data of the highest value (i.e. highest hash of
+   the highest seqno).
+4. A replay set is constructed; this is ordered and must contain some reference to the message that
+   contains the diff (so that added/changed values can have their value retrieved from the
+   appropriate message).
+   - the "current" diff of every message is added to the replay sets (the seqno and hash are
+     extracted and computed from the message itself).
+   - past diffs in the messages (i.e. from the `"<"` list) are added to the replay set, *if* they
+     are not yet in the reply set, and have a seqno >= newseqno-5.  This should be done in order
```

Comment:
typo: `reply set` should be `replay set`

Reply:
Thanks, fixed.